src/librustc/ty/sty.rs

```diff
@@ -254,15 +257,38 @@ pub enum TypeVariants<'tcx> {
 /// handle). Plus it fixes an ICE. :P
 #[derive(Copy, Clone, PartialEq, Eq, Hash, Debug, RustcEncodable, RustcDecodable)]
 pub struct ClosureSubsts<'tcx> {
-    /// Lifetime and type parameters from the enclosing function.
+    /// Lifetime and type parameters from the enclosing function,
+    /// concatenated with the types of the upvars.
+    ///
     /// These are separated out because trans wants to pass them around
     /// when monomorphizing.
-    pub func_substs: &'tcx Substs<'tcx>,
+    pub substs: &'tcx Substs<'tcx>,
+}
 
-    /// The types of the upvars. The list parallels the freevars and
-    /// `upvar_borrows` lists. These are kept distinct so that we can
-    /// easily index into them.
-    pub upvar_tys: &'tcx Slice<Ty<'tcx>>
+// using a real closure here results in lifetime capture errors, so
+// do it manually.
+
+pub struct KindAsType;
+impl<'tcx> FnOnce<(&'tcx Kind<'tcx>,)> for KindAsType {
+    type Output = Ty<'tcx>;
+    extern "rust-call" fn call_once(mut self, args: (&'tcx Kind<'tcx>,)) -> Ty<'tcx> {
+        self.call_mut(args)
+    }
+}
+impl<'tcx> FnMut<(&'tcx Kind<'tcx>,)> for KindAsType {
+    extern "rust-call" fn call_mut(&mut self, args: (&'tcx Kind<'tcx>,)) -> Ty<'tcx> {
+        args.0.as_type().expect("unexpected region in upvars")
+    }
+}
+
+impl<'a, 'gcx, 'acx, 'tcx> ClosureSubsts<'tcx> {
+    #[inline]
+    pub fn upvar_tys(self, def_id: DefId, tcx: TyCtxt<'a, 'gcx, 'acx>) ->
+        iter::Map<slice::Iter<'tcx, Kind<'tcx>>, KindAsType>
```

Comment:
Should just use `impl Trait` here, maybe you can share the error?
Last I checked most cases can be tricked into inferring correctly.